File: g2o/examples/icp/gicp_sba_demo.cpp
// g2o - General Graph Optimization
// Copyright (C) 2011 Kurt Konolige
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright
//   notice, this list of conditions and the following disclaimer in the
//   documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <cstdint>
#include <iostream>
#include <random>

#include "g2o/core/block_solver.h"
#include "g2o/core/optimization_algorithm_levenberg.h"
#include "g2o/core/solver.h"
#include "g2o/core/sparse_optimizer.h"
#include "g2o/solvers/eigen/linear_solver_eigen.h"
#include "g2o/stuff/sampler.h"
#include "g2o/types/icp/types_icp.h"

using std::cerr;
using std::cout;
using std::endl;

namespace g2o {
static int gicp_sba_demo(int argc, char** argv) {
  int num_points = 0;

  // check for arg, # of points to use in projection SBA
  if (argc > 1) num_points = atoi(argv[1]);

  double euc_noise = 0.1;  // noise in position, m
  double pix_noise = 1.0;  // pixel noise
  //  double outlier_ratio = 0.1;

  SparseOptimizer optimizer;
  optimizer.setVerbose(false);

  // variable-size block solver
  g2o::OptimizationAlgorithm* solver =
      new g2o::OptimizationAlgorithmLevenberg(g2o::make_unique<BlockSolverX>(
          g2o::make_unique<
              LinearSolverEigen<g2o::BlockSolverX::PoseMatrixType>>()));

  optimizer.setAlgorithm(std::unique_ptr<OptimizationAlgorithm>(solver));

  std::vector<Vector3> true_points;
  for (size_t i = 0; i < 1000; ++i) {
    true_points.emplace_back((g2o::Sampler::uniformRand(0., 1.) - 0.5) * 3,
                             g2o::Sampler::uniformRand(0., 1.) - 0.5,
                             g2o::Sampler::uniformRand(0., 1.) + 10);
  }

  // set up camera params
  Vector2 focal_length(500, 500);     // pixels
  Vector2 principal_point(320, 240);  // 640x480 image
  double baseline = 0.075;            // 7.5 cm baseline

  // set up camera params and projection matrices on vertices
  g2o::VertexSCam::setKcam(focal_length[0], focal_length[1], principal_point[0],
                           principal_point[1], baseline);

  // set up two poses
  int vertex_id = 0;
  for (size_t i = 0; i < 2; ++i) {
    // set up rotation and translation for this node
    Vector3 t(0, 0, i);
    Quaternion q;
    q.setIdentity();

    Eigen::Isometry3d cam;  // camera pose
    cam = q;
    cam.translation() = t;

    // set up node
    auto vc = std::make_shared<VertexSCam>();
    vc->setEstimate(cam);
    vc->setId(vertex_id);  // vertex id

    cerr << t.transpose() << " | " << q.coeffs().transpose() << endl;

    // set first cam pose fixed
    if (i == 0) vc->setFixed(true);

    // make sure projection matrices are set
    vc->setAll();

    // add to optimizer
    optimizer.addVertex(vc);

    vertex_id++;
  }

  // set up point matches for GICP
  for (size_t i = 0; i < true_points.size(); ++i) {
    // get two poses
    auto vp0 = std::dynamic_pointer_cast<VertexSE3>(
        optimizer.vertices().find(0)->second);
    auto vp1 = std::dynamic_pointer_cast<VertexSE3>(
        optimizer.vertices().find(1)->second);

    // calculate the relative 3D position of the point
    Vector3 pt0 = vp0->estimate().inverse() * true_points[i];
    Vector3 pt1 = vp1->estimate().inverse() * true_points[i];

    // add in noise
    pt0 += Vector3(g2o::Sampler::gaussRand(0., euc_noise),
                   g2o::Sampler::gaussRand(0., euc_noise),
                   g2o::Sampler::gaussRand(0., euc_noise));

    pt1 += Vector3(g2o::Sampler::gaussRand(0., euc_noise),
                   g2o::Sampler::gaussRand(0., euc_noise),
                   g2o::Sampler::gaussRand(0., euc_noise));

    // form edge, with normals in varioius positions
    Vector3 nm0(0, i, 1);
    Vector3 nm1(0, i, 1);
    nm0.normalize();
    nm1.normalize();

    auto e  // new edge with correct cohort for caching
        = std::make_shared<EdgeVVGicp>();

    e->vertices()[0]  // first viewpoint
        = vp0;

    e->vertices()[1]  // second viewpoint
        = vp1;

    EdgeGICP meas;

    meas.pos0 = pt0;
    meas.pos1 = pt1;
    meas.normal0 = nm0;
    meas.normal1 = nm1;
    e->setMeasurement(meas);
    meas = e->measurement();
    //        e->inverseMeasurement().pos() = -kp;

    // use this for point-plane
    e->information() = meas.prec0(0.01);

    // use this for point-point
    //    e->information().setIdentity();

    //    e->setRobustKernel(true);
    // e->setHuberWidth(0.01);

    optimizer.addEdge(e);
  }

  // set up SBA projections with some number of points

  true_points.clear();
  for (int i = 0; i < num_points; ++i) {
    true_points.emplace_back((g2o::Sampler::uniformRand(0., 1.) - 0.5) * 3,
                             g2o::Sampler::uniformRand(0., 1.) - 0.5,
                             g2o::Sampler::uniformRand(0., 1.) + 10);
  }

  // add point projections to this vertex
  for (auto& true_point : true_points) {
    auto v_p = std::make_shared<g2o::VertexPointXYZ>();

    v_p->setId(vertex_id++);
    v_p->setMarginalized(true);
    v_p->setEstimate(true_point + Vector3(g2o::Sampler::gaussRand(0., 1),
                                          g2o::Sampler::gaussRand(0., 1),
                                          g2o::Sampler::gaussRand(0., 1)));

    optimizer.addVertex(v_p);

    for (size_t j = 0; j < 2; ++j) {
      Vector3 z;
      std::dynamic_pointer_cast<g2o::VertexSCam>(
          optimizer.vertices().find(j)->second)
          ->mapPoint(z, true_point);

      if (z[0] >= 0 && z[1] >= 0 && z[0] < 640 && z[1] < 480) {
        z += Vector3(g2o::Sampler::gaussRand(0., pix_noise),
                     g2o::Sampler::gaussRand(0., pix_noise),
                     g2o::Sampler::gaussRand(0., pix_noise / 16.0));

        auto e = std::make_shared<g2o::EdgeXyzVsc>();

        e->vertices()[0] = v_p;

        e->vertices()[1] = optimizer.vertices().find(j)->second;

        e->setMeasurement(z);
        // e->inverseMeasurement() = -z;
        e->information() = Matrix3::Identity();

        // e->setRobustKernel(false);
        // e->setHuberWidth(1);

        optimizer.addEdge(e);
      }
    }
  }  // done with adding projection points

  // move second cam off of its true position
  auto vc = std::dynamic_pointer_cast<VertexSE3>(
      optimizer.vertices().find(1)->second);
  Eigen::Isometry3d cam = vc->estimate();
  cam.translation() = Vector3(-0.1, 0.1, 0.2);
  vc->setEstimate(cam);
  optimizer.initializeOptimization();
  optimizer.computeActiveErrors();
  cout << "Initial chi2 = " << FIXED(optimizer.chi2()) << endl;

  optimizer.setVerbose(true);

  optimizer.optimize(20);

  cout << endl << "Second vertex should be near 0,0,1" << endl;
  cout << std::dynamic_pointer_cast<VertexSE3>(
              optimizer.vertices().find(0)->second)
              ->estimate()
              .translation()
              .transpose()
       << endl;
  cout << std::dynamic_pointer_cast<VertexSE3>(
              optimizer.vertices().find(1)->second)
              ->estimate()
              .translation()
              .transpose()
       << endl;
  return 0;
}
}  // namespace g2o

int main(int argc, char** argv) { return g2o::gicp_sba_demo(argc, argv); }

File: python/examples/gicp_sba_demo.py
# https://github.com/RainerKuemmerle/g2o/blob/master/g2o/examples/icp/gicp_sba_demo.cpp

import numpy as np 
import g2o

import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--num_points', type=int, help='num of points to use in projection SBA', default=0)
parser.add_argument('--pos_noise', type=float, help='noise in 3d position', default=0.2)
parser.add_argument('--pixel_noise', type=float, help='pixel noise', default=0.5)
parser.add_argument('--seed', type=int, help='random seed', default=0)
args = parser.parse_args()



def main():
    optimizer = g2o.SparseOptimizer()
    solver = g2o.BlockSolverX(g2o.LinearSolverCSparseX())
    solver = g2o.OptimizationAlgorithmLevenberg(solver)
    optimizer.set_algorithm(solver)

    true_points = np.hstack([
        np.random.random((1000, 1)) * 3 - 1.5,
        np.random.random((1000, 1)) - 0.5,
        np.random.random((1000, 1)) + 10])

    
    focal_length = (500, 500)
    principal_point = (320, 240)
    baseline = 0.075
    g2o.VertexSCam.set_cam(*focal_length, *principal_point, baseline)

    for i in range(2):
        t = np.array([0, 0, i])
        cam = g2o.Isometry3d(np.identity(3), t)

        vc = g2o.VertexSCam()
        vc.set_id(i)
        vc.set_estimate(cam)
        if i == 0:
            vc.set_fixed(True)
        vc.set_all()
        optimizer.add_vertex(vc)


    trans0 = optimizer.vertex(0).estimate().inverse()
    trans1 = optimizer.vertex(1).estimate().inverse()

    for i in range(len(true_points)):
        pt0 = trans0 * true_points[i]
        pt1 = trans1 * true_points[i]

        # add noise
        pt0 += np.random.randn(3) * args.pos_noise
        pt1 += np.random.randn(3) * args.pos_noise

        # form edge, with normals in varioius positions
        nm0 = np.array([0, i, 1])
        nm0 = nm0 / np.linalg.norm(nm0)
        nm1 = np.array([0, i, 1])
        nm1 = nm1 / np.linalg.norm(nm1)

        meas = g2o.EdgeGICP()
        meas.pos0 = pt0
        meas.pos1 = pt1
        meas.normal0 = nm0
        meas.normal1 = nm1

        edge = g2o.Edge_V_V_GICP()
        edge.set_vertex(0, optimizer.vertex(0))
        edge.set_vertex(1, optimizer.vertex(1))
        edge.set_measurement(meas)
        edge.set_information(meas.prec0(0.01))

        optimizer.add_edge(edge)


    # set up SBA projections
    if args.num_points > 0:
        true_points = np.hstack([
            np.random.random((args.num_points, 1)) * 3 - 1.5,
            np.random.random((args.num_points, 1)) - 0.5,
            np.random.random((args.num_points, 1)) + 10])

        cam_num = 2
        for i, point in enumerate(true_points):
            vp = g2o.VertexSBAPointXYZ()
            vp.set_id(cam_num + i)
            vp.set_marginalized(True)
            vp.set_estimate(point + np.random.randn(3))
            optimizer.add_vertex(vp)

            for j in range(cam_num):
                z = optimizer.vertex(j).map_point(point)
                if 0 <= z[0] < 640 and 0 <= z[1] < 480:
                    z += np.random.randn(3) * args.pixel_noise * [1, 1, 1/16.]

                    edge = g2o.Edge_XYZ_VSC()
                    edge.set_vertex(0, vp)
                    edge.set_vertex(1, optimizer.vertex(j))
                    edge.set_measurement(z)
                    edge.set_information(np.identity(3))
                    edge.set_robust_kernel(g2o.RobustKernelHuber())

                    optimizer.add_edge(edge)


    # move second cam off of its true position
    vc = optimizer.vertex(1)
    cam = g2o.Isometry3d(vc.estimate().R, np.array([-0.1, -0.1, 0.2]))
    vc.set_estimate(cam)


    optimizer.initialize_optimization()
    optimizer.compute_active_errors()
    print('Initial chi2 =', optimizer.chi2())

    optimizer.set_verbose(True)
    optimizer.optimize(20)

    print('\nSecond vertex should be near [0, 0, 1]')
    print('before optimization:', cam.t)
    print('after  optimization:', optimizer.vertex(1).estimate().t)
    print('error:', optimizer.vertex(1).estimate().t - [0, 0, 1])


    '''
    Mean squared error (average over 100 loops):
    num_points     0:   [ 0.29733384  0.40814327  0.03907623]
                  10:   [ 0.14172112  0.31366953  0.03670497]
                 100:   [ 0.05319327  0.37413272  0.01637925]
                1000:   [ 0.01652139  0.19501433  0.00502872]
               10000:   [ 0.00901297  0.04055765  0.00130791]   ->  1/3 : 1 : 1/16
    
    '''



if __name__ == '__main__':
    if args.seed > 0:
        np.random.seed(args.seed)

    main()